### PR TITLE
Introduce Scalar, SeriesLine, and SeriesPoint archetypes with their own visualizers

### PR DESCRIPTION
**Pull request by jleibs**
### What
- New Scalar, SeriesLine, and SeriesPoint types
- Factor out common plot operations for reuse across visualizers
- New visualizer systems for Line and Point
- New component to control overriden visualizers
- New UI to modify the component

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4875/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4875/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4875/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4875)
- [Docs preview](https://rerun.io/preview/ff87c9e817d75c26ad596ac678378c41b35b9d23/docs) 
- [Examples preview](https://rerun.io/preview/ff87c9e817d75c26ad596ac678378c41b35b9d23/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_space_view_time_series/src/aggregation.rs
Comment (jleibs) [inline crates/re_space_view_time_series/src/aggregation.rs:1]: This is just moving all the aggregation pieces out of visualizer_system.

File: crates/re_space_view_time_series/src/util.rs
Comment (jleibs) [inline crates/re_space_view_time_series/src/util.rs:1]: This is moving more common pieces out of visualizer_system.

File: crates/re_types/definitions/rerun/archetypes/scalar.fbs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace rerun.archetypes;
+
+// ---
+
+/// Log a double-precision scalar.
+///
+/// The current simulation time will be used for the time/X-axis, hence scalars
+/// cannot be timeless!
+///
```
Comment (Wumpf) [inline crates/re_types/definitions/rerun/archetypes/scalar.fbs:14]: I think it would make sense to refer to the "related" archetypes of `series_line` and `series_point`. Since it confused users in the past I think it's also important to point out that we expect you to log the same ones on the same path

File: crates/re_types/definitions/rerun/blueprint/archetypes/series_line.fbs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+include "fbs/attributes.fbs";
+include "rust/attributes.fbs";
+
+include "rerun/components.fbs";
+
+namespace rerun.archetypes;
```
Comment (Wumpf) [inline crates/re_types/definitions/rerun/blueprint/archetypes/series_line.fbs:6]: would be still nice to move this (and _points) to the reguilar archetype directory. A bit confusing ;)

File: crates/re_types/definitions/rerun/components/visualizer_overrides.fbs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// ---
+
+/// The name of a visualizer.
+table VisualizerOverrides (
+    "attr.docs.unreleased",
+    "attr.rust.derive": "PartialEq, Eq, PartialOrd, Ord",
+    "attr.rust.repr": "transparent",
+    "attr.rust.override_crate": "re_types_core"
+) {
+    // TODO(jleibs): Can we do this with numeric type-ids instead?
```
Comment (Wumpf) [inline crates/re_types/definitions/rerun/components/visualizer_overrides.fbs:19]: ... and would we want to :thinking:

File: crates/re_space_view/src/data_query_blueprint.rs [hunk 4]
```diff
@@ -467,6 +470,20 @@ impl DataQueryPropertyResolver<'_> {
                         .individual_override_root
                         .join(&node.data_result.entity_path);
 
+                    // If the user has overridden the visualizers, update which visualizers are used.
+                    {
```
Comment (Wumpf) [inline crates/re_space_view/src/data_query_blueprint.rs:474]: mesmerizing how little code does the actual lifting of filtering out visualizers :)
should we check against opf --no-frames if this extra query does us any perf harm?

File: crates/re_space_view/src/data_query_blueprint.rs [hunk 5]
```diff
@@ -254,6 +254,9 @@ impl<'a> QueryExpressionEvaluator<'a> {
 
         // Only populate visualizers if this is a match
         // Note that allowed prefixes that aren't matches can still create groups
+        // TODO(jleibs): It would be nice to lookup the override queries here, but we don't have
+        // access to a query context. Also the entity-override-path-joining is expensive and we don't want
+        // to do it during heuristic evaluation.
```
Comment (Wumpf) [inline crates/re_space_view/src/data_query_blueprint.rs:259]: would be nice to refer to the other place where we do it instead

File: crates/re_space_view_time_series/src/space_view_class.rs [hunk 6]
```diff
@@ -210,7 +216,7 @@ It can greatly improve performance (and readability) in such situations as it pr
         // For all following lookups, checking indicators is enough, since we know that this is enough to infer visualizability here.
         let Some(indicated_entities) = ctx
             .indicated_entities_per_visualizer
-            .get(&TimeSeriesSystem::identifier())
+            .get(&LegacyTimeSeriesSystem::identifier())
         else {
```
Comment (Wumpf) [inline crates/re_space_view_time_series/src/space_view_class.rs:220]: Need to patch this: This should take the union of Legacy, Point and Line systems

File: crates/re_space_view_time_series/src/space_view_class.rs [hunk 7]
```diff
@@ -286,13 +292,33 @@ It can greatly improve performance (and readability) in such situations as it pr
...
         let timeline_name = timeline.name().to_string();
 
-        let time_series = system_output.view_systems.get::<TimeSeriesSystem>()?;
+        let legacy_time_series = system_output.view_systems.get::<LegacyTimeSeriesSystem>()?;
+        let line_series = system_output.view_systems.get::<SeriesLineSystem>()?;
+        let point_series = system_output.view_systems.get::<SeriesPointSystem>()?;
 
-        let aggregator = time_series.aggregator;
-        let aggregation_factor = time_series.aggregation_factor;
+        let all_lines: Vec<_> = legacy_time_series
```
Comment (Wumpf) [inline crates/re_space_view_time_series/src/space_view_class.rs:299]:
```suggestion
        let all_plot_items: Vec<_> = legacy_time_series
```

It's not just lines!

File: crates/re_space_view_time_series/src/space_view_class.rs [hunk 8]
```diff
@@ -373,7 +399,7 @@ It can greatly improve performance (and readability) in such situations as it pr
                 time_ctrl_write.pause();
             }
 
-            for line in &time_series.lines {
+            for line in all_lines {
```
Comment (Wumpf) [inline crates/re_space_view_time_series/src/space_view_class.rs:402]: for whatever reason this was called lines before, it's super confusing now that I review stuff here

File: crates/re_space_view_time_series/src/visualizer_system.rs [hunk 9]
```diff
@@ -1,91 +1,34 @@
...
+    overrides::{initial_override_color, lookup_override},
+    util::{determine_time_range, points_to_lines},
+    PlotPoint, PlotPointAttrs, PlotSeries, PlotSeriesKind,
+};
 
-/// A scene for a time series plot, with everything needed to render it.
+/// The legacy system for rendering [`TimeSeriesScalars`] archetypes.
 #[derive(Default, Debug)]
-pub struct TimeSeriesSystem {
+pub struct LegacyTimeSeriesSystem {
```
Comment (Wumpf) [inline crates/re_space_view_time_series/src/visualizer_system.rs:20]: rename the file as well?

File: crates/re_space_view_time_series/src/visualizer_system.rs [hunk 10]
```diff
@@ -1,91 +1,34 @@
...
     fn identifier() -> re_viewer_context::ViewSystemIdentifier {
-        "TimeSeries".into()
+        "LegacyTimeSeries".into()
     }
 }
 
-impl VisualizerSystem for TimeSeriesSystem {
+impl VisualizerSystem for LegacyTimeSeriesSystem {
     fn visualizer_query_info(&self) -> VisualizerQueryInfo {
         VisualizerQueryInfo::from_archetype::<TimeSeriesScalar>()
```
Comment (Wumpf) [inline crates/re_space_view_time_series/src/visualizer_system.rs:33]: we should still leave some note on `TimeSeriesScalar` that it's on the way out. Otherwise, we always end up with the legacy time series system which might be confusing ("why is it saying legacy no matter what I do")

File: crates/re_space_view_time_series/src/lib.rs [hunk 11]
```diff
@@ -2,9 +2,16 @@
 //!
 //! A Space View that shows plots over Rerun timelines.
 
+mod aggregation;
+mod legacy_visualizer_system;
```
Comment (Wumpf) [inline crates/re_space_view_time_series/src/lib.rs:6]: I'm clearly missing a file or a rename, this file is not checked in yet!

File: crates/re_space_view_time_series/src/line_visualizer_system.rs [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+        let egui_ctx = &ctx.re_ui.egui_ctx;
+
+        let plot_mem = egui_plot::PlotMemory::load(egui_ctx, crate::plot_id(query.space_view_id));
+        let plot_bounds = plot_mem.as_ref().map(|mem| *mem.bounds());
+        // What's the delta in value of X across two adjacent UI points?
+        // I.e. think of GLSL's `dpdx()`.
+        let plot_value_delta = plot_mem.as_ref().map_or(1.0, |mem| {
+            1.0 / mem.transform().dpos_dvalue_x().max(f64::EPSILON)
+        });
```
Comment (Wumpf) [inline crates/re_space_view_time_series/src/line_visualizer_system.rs:102]: this is the code that determines the aggregation factor and as such I'm not super happy having it triplicated even if the critical part is relatively short. Can we further integrate this between the visualizers?

**Review by Wumpf — CHANGES_REQUESTED** [review]
looking good overall. Heuristic needs patching and duplication of amount of aggregation is problematic

Was fairly trusting on the moved code, hard to follow all of the details how which thing moved

File: crates/re_space_view_time_series/src/lib.rs (hunk 11 above)
Comment (jleibs) [reply crates/re_space_view_time_series/src/lib.rs]: Oops, yeah. I named it back because it made the diff much easier to review but messed up the last pus.

File: crates/re_space_view_time_series/src/line_visualizer_system.rs (hunk 12 above)
Comment (jleibs) [reply crates/re_space_view_time_series/src/line_visualizer_system.rs]: good call

File: crates/re_types/definitions/rerun/archetypes/scalar.fbs [hunk 13]
```diff
@@ -9,8 +9,16 @@ namespace rerun.archetypes;
 
 /// Log a double-precision scalar.
 ///
-/// The current simulation time will be used for the time/X-axis, hence scalars
-/// cannot be timeless!
+/// The current timeline value will be used for the time/X-axis, hence scalars
+/// cannot be timeless.
+///
+/// WHen used to produce a plot, this archetytpe is used to provide the data that
```
Comment (Wumpf) [inline crates/re_types/definitions/rerun/archetypes/scalar.fbs:15]:
```suggestion
/// When used to produce a plot, this archetytpe is used to provide the data that
```

File: crates/re_space_view_time_series/src/visualizer_system.rs (hunk 9 above)
Comment (jleibs) [reply crates/re_space_view_time_series/src/visualizer_system.rs:20]: I originally did but it made the git diff much more annoying because it failed to track the move

File: crates/re_space_view_time_series/src/space_view_class.rs [hunk 14]
```diff
@@ -214,12 +214,22 @@ It can greatly improve performance (and readability) in such situations as it pr
...
-            .indicated_entities_per_visualizer
-            .get(&LegacyTimeSeriesSystem::identifier())
-        else {
+        let mut indicated_entities = IndicatedEntities::default();
+
+        for indicated in [
+            LegacyTimeSeriesSystem::identifier(),
+            SeriesLineSystem::identifier(),
+            SeriesPointSystem::identifier(),
+        ]
```
Comment (Wumpf) [inline crates/re_space_view_time_series/src/space_view_class.rs:223]: Hmm if one would only log the new scalar archetype, then this wouldn't spawn a space view. But then again we also don't know what visualizer to activate right now, so I figure this is correct? All a bit confusing with the Legacy one around right now

File: tests/roundtrips.py [hunk 15]
```diff
@@ -28,6 +28,9 @@
     "clear": ["cpp", "py", "rust"],  # Don't need it, API example roundtrips cover it all
     "mesh3d": ["cpp", "py", "rust"],  # Don't need it, API example roundtrips cover it all
     "time_series_scalar": ["cpp", "py", "rust"],  # Don't need it, API example roundtrips cover it all
+    "scalar": ["cpp", "py", "rust"],  # TODO(jleibs)
+    "series_line": ["cpp", "py", "rust"],  # TODO(jleibs)
+    "series_point": ["cpp", "py", "rust"],  # TODO(jleibs)
```
Comment (Wumpf) [inline tests/roundtrips.py:33]: do we _really_ need roundtrip tests for these? worth bringing up as a post-standup topic how many of these we want/need in the future

**Review by Wumpf — APPROVED** [review]
🚢

File: crates/re_space_view_time_series/src/space_view_class.rs (hunk 14 above)
Comment (jleibs) [reply crates/re_space_view_time_series/src/space_view_class.rs:223]: This actually works because the SeriesLineSystem and SeriesPointSystem both accept Scalar as an indicator.

File: tests/roundtrips.py (hunk 15 above)
Comment (jleibs) [reply tests/roundtrips.py:33]: yeah, TODO there until I've at least made sure they are covered somehow. With it's implicit via code-examples or explicit via RTT.
